File: Level of Nodes - GFG/level-of-nodes.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution
{
	public:
	//Function to find the level of node X.
	int nodeLevel(int n, vector<int> adj[], int x) 
	{
	    vector <int> vis(n+1);
	    queue <int> q;
	    q.push(0);
	    int lev = 0;
	    if (x == 0) return 0;
	    vis[0] = 1;
	    while(q.size())
	    {
	        int len = q.size();
	        for (int i = 0; i < len; i++)
	        {
	            int node = q.front();
	            q.pop();
	            for (int child : adj[node])
	            {
	                if (!vis[child]) {
    	                if (child == x) return lev+1;
    	                q.push(child);
    	                vis[child] = 1;
	                }
	            }
	        }
	        lev++;
	    }
	    return -1;
	}
};

//{ Driver Code Starts.


int main()
{
    
    int t;
    cin >> t;
    while(t--)
    {
    	int V, E, X;
    	cin >> V >> E;

    	vector<int> adj[V];

    	for(int i = 0; i < E; i++)
    	{
    		int u, v;
    		cin >> u >> v;
    		adj[u].push_back(v);
    		adj[v].push_back(u);
    	}
    	cin>>X;

    	Solution obj;
    	cout << obj.nodeLevel(V, adj, X) << "\n";
    }

    return 0;
}


// } Driver Code Ends

File: Level of Nodes - GFG/level-of-nodes.cs
//{ Driver Code Starts
//Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DriverCode
{

    class GFG
    {
        static void Main(string[] args)
        {
            int testcases;// Taking testcase as input
            testcases = Convert.ToInt32(Console.ReadLine());
            while (testcases-- > 0)// Looping through all testcases
            {
                var ip = Console.ReadLine().Trim().Split(' ');
                int V = int.Parse(ip[0]);
                int E = int.Parse(ip[1]);
                List<List<int>> adj = new List<List<int>>();
                for (int i = 0; i < V; i++)
                {
                    adj.Add(new List<int>());
                }
                for (int i = 0; i < E; i++)
                {
                    ip = Console.ReadLine().Trim().Split(' ');
                    int u = int.Parse(ip[0]);
                    int v = int.Parse(ip[1]);
                    adj[u].Add(v);
                    adj[v].Add(u);
                }
                int X = Convert.ToInt32(Console.ReadLine());
                Solution obj = new Solution();
                var res = obj.nodeLevel(V, adj, X);
                Console.WriteLine(res);
            }
        }
    }
}
// } Driver Code Ends


//User function Template for C#

class Solution
{
    public void dfs(int node, List<List<int>> adj, int[] dist, int[] vis)
	{
	    vis[node] = 1;
        foreach (int child in adj[node])
        {
            if (vis[child] == 0)
            {
                dist[child] = dist[node] + 1;
                dfs(child, adj, dist, vis);
            }
        }
	}
    public int nodeLevel(int n, List<List<int>> adj, int x)
    {
        int[] dist = new int[n+1];
        int[] vis = new int[n+1];

        dfs(0, adj, dist, vis);

        if (x == 0) return 0;
        if (dist[x] == 0) return -1;
        return dist[x];
    }
}